src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.SymbolStartAnalyzer.BlockAnalyzer.cs

```diff
@@ -0,0 +1,490 @@
+
+                        if (isUnusedLocalAssignment &&
+                            !isRemovableAssignment &&
+                            _options.UnusedValueAssignmentPreference == UnusedValuePreference.UnusedLocalVariable)
+                        {
+                            // Meets current user preference of using unused local symbols for storing computation result.
+                            // Skip reporting diagnostic.
+                            return false;
+                        }
+
+                        properties = s_propertiesMap[(_options.UnusedValueAssignmentPreference, isUnusedLocalAssignment, isRemovableAssignment)];
+                        return true;
+                    }
+
+                    // Indicates if the given unused symbol write is a removable assignment.
+                    // This is true if the expression for the assigned value has no side effects.
+                    bool IsRemovableAssignmentWithoutSideEffects(IOperation unusedSymbolWriteOperation)
+                    {
+                        if (unusedSymbolWriteOperation.Parent is IAssignmentOperation assignment &&
+                            assignment.Target == unusedSymbolWriteOperation)
+                        {
+                            if (assignment.Value.ConstantValue.HasValue)
+                            {
+                                // Constant expressions have no side effects.
+                                return true;
+                            }
+
+                            switch (assignment.Value.Kind)
+                            {
+                                // Parameter/local references have no side effects and can be removed.
+                                case OperationKind.ParameterReference:
+                                case OperationKind.LocalReference:
+                                    return true;
+
+                                // Field references with null instance (static fields) or 'this' or 'Me' instance can
+                                // have no side effects and can be removed.
+                                case OperationKind.FieldReference:
+                                    var fieldReference = (IFieldReferenceOperation)assignment.Value;
+                                    return fieldReference.Instance == null || fieldReference.Instance.Kind == OperationKind.InstanceReference;
+                            }
+                        }
+                        else if (unusedSymbolWriteOperation.Parent is IIncrementOrDecrementOperation)
+                        {
+                            // As the new value assigned to the incremented/decremented variable is unused,
+                            // it is safe to remove the entire increment/decrement operation,
+                            // as it cannot have side effects on anything but the variable.
+                            return true;
```

Comment:
note: in a language like C this might not be true if the value was volatile.  I don't believe volatile in .net means the same thing (i.e. 'may be backed by a system that can observe mutations').  But we may want to be careful here.  Can you open a bug to investigate implications of eliding writes on volatile fields?

Reply:
Also, is this safe for pre/post increment?  If this is a post-increment, is the value assigned the value *prior* to the increment?

Reply:
This analyzer and the underlying SymbolUsageAnalysis only operates on locals and parameters.

---
In reply to: [236088420](https://github.com/dotnet/roslyn/pull/30777#discussion_r236088420) [](ancestors = 236088420)

Reply:
This identifies `i++;` or `++i` on a local/parameter, such that new value of `i` is never read. Such increment/decrement expressions can be completely removed.

---
In reply to: [236088462](https://github.com/dotnet/roslyn/pull/30777#discussion_r236088462) [](ancestors = 236088462)